### PR TITLE
Support SSO 

### LINKED ISSUE #190: Support SSO
### ⚠️ Please check that this feature request hasn't been suggested before.

- [X] I searched previous [Ideas in Discussions](https://github.com/homanp/superagent/discussions/categories/ideas) didn't find any similar feature requests.
- [X] I searched previous [Issues](https://github.com/homanp/superagent/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

Superagent should support SSO

### 📝 Additional Context

_No response_

### Acknowledgements

- [X] My issue title is concise and descriptive.
- [X] I have searched the existing issues to make sure this feature has not been requested yet.
- [X] I have provided enough information for the maintainers to understand and evaluate this request.

### PR DESCRIPTION
**Pull request by ankushbhardwxj**
## Summary

This PR adds support for authentication via SSO. 

Fixes #190 

## Checklist

- [x] My code follows the code style of this project and passes `make format`.
- [ ] My change requires a change to the documentation.
- [ ] I have [updated the documentation](https://docs.superagent.sh/) accordingly.
- [ ] My change has adequate unit test coverage.

### CODE REVIEW

File: ui/lib/next-auth.js [hunk 1]
```diff
@@ -29,6 +32,19 @@ export const options = {
...
+    }),
+    GoogleProvider({
+      clientId: process.env.GOOGLE_CLIENT_ID,
+      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
+    }),
+    AzureADProvider({
+      clientId: process.env.AZURE_AD_CLIENT_ID,
+      clientSecret: process.env.AZURE_AD_CLIENT_SECRET,
+      tenantId: process.env.AZURE_AD_TENANT_ID,
+    }),
```
Comment (ankushbhardwxj) [inline ui/lib/next-auth.js:47]: @homanp : Require these client ids & secrets for the providers.

Comment (ankushbhardwxj) [reply ui/lib/next-auth.js:47]: I believe AzureADProvider is same as Microsoft Provider. NextAuth doesn't have specifically by the name of Microsoft.

Comment (homanp) [reply ui/lib/next-auth.js:47]: Correct!

Comment (homanp) [reply ui/lib/next-auth.js:47]: Please ping when you are ready with this PR.

Comment (homanp) [reply ui/lib/next-auth.js:47]: @ankushbhardwxj do you need any help with this? If so, ping.

Comment (ankushbhardwxj) [reply ui/lib/next-auth.js:47]: Didn't get time to get to it yesterday. Will try to wrap this up today. Will ping, if help needed.

**Comment by ankushbhardwxj** [comment]
@homanp : I guess this PR is ready for a review. 

Two questions:
- Am I supposed to push my `yarn.lock` file or db migration files ?
- Didn't implement Refresh Token flow as of now. Do we want to support that ? I guess only Google & Microsoft have refresh token flow.

Do let me know for any changes 👍🏻

**Comment by homanp** [comment]
> @homanp : I guess this PR is ready for a review.
> 
> Two questions:
> 
> * Am I supposed to push my `yarn.lock` file, db migration files and `.env` files ?
> * Didn't implement Refresh Token flow as of now. Do we want to support that ? I guess only Google & Microsoft have refresh token flow.
> 
> Do let me know for any changes 👍🏻

Hey, from a glance this looks really good. 
1. You should add your migration files. 
2. No need to add `.env` or `yarn.lock`. 
3. Regarding the refresh token, what happens if we don’t implement, the user gets logged out?

Also did you have the chance to test signup/login with your changes? Everything works?

**Comment by ankushbhardwxj** [comment]
> Also did you have the chance to test signup/login with your changes? Everything works?

Yes, sign up and login works fine. 

> Regarding the refresh token, what happens if we don’t implement, the user gets logged out?

Yes, the user will require to log in again. Frequency of that depends on expiration duration of the provider we are working with. 

Haven't implemented the authentication middleware yet on the backend. Basically, that's where I'm planning to refresh user tokens. We can also refresh the token on next-auth side. Do let me know which flow seems suitable to you.

**Comment by homanp** [comment]
> > Also did you have the chance to test signup/login with your changes? Everything works?
> 
> Yes, sign up and login works fine.
> 
> > Regarding the refresh token, what happens if we don’t implement, the user gets logged out?
> 
> Yes, the user will require to log in again. Frequency of that depends on expiration duration of the provider we are working with.
> 
> Haven't implemented the authentication middleware yet on the backend. Basically, that's where I'm planning to refresh user tokens. We can also refresh the token on next-auth side. Do let me know which flow seems suitable to you.

@ankushbhardwxj we can leave it as is. Could we make the different OAuth options optional? This would make it easier for people fork. Could you add that in as well?

**Comment by ankushbhardwxj** [comment]
> Could we make the different OAuth options optional?

I think they are already optional. If somebody doesn't want to use SSO, they can use the login form.

**Review by homanp — CHANGES_REQUESTED** [review]
@ankushbhardwxj I noticed two issues when testing: 

1) Regular credentials stopped working (this is fixed now by me)
2) Can't create any objects due to the authorization token missing in all requests.

Can you please fix point 2.

**Comment by ankushbhardwxj** [comment]
On testing, found another issue - If OAuth was done to log in first, user cannot sign up using credentials. User should be able to use both flows.

**Comment by homanp** [comment]
> On testing, found another issue - If OAuth was done to log in first, user cannot sign up using credentials. User should be able to use both flows. 

Could you fix?

**Comment by ankushbhardwxj** [comment]
I'm onto it, shall push in some time

**Comment by homanp** [comment]
> I'm onto it, shall push in some time. Also, fixing 2 from above

Isn't the user signed up when using oauth the first time? Why would they need to signup again?

**Comment by ankushbhardwxj** [comment]
> Isn't the user signed up when using oauth the first time? Why would they need to signup again?

I have found instances, when users although they signed up using Credentials the first time, tried to use log in with OAuth later and vice versa. 
However, it's not super important to support it, we can drop it if required.

**Comment by homanp** [comment]
> > Isn't the user signed up when using oauth the first time? Why would they need to signup again?
> 
> 
> 
> I have found instances, when users although they signed up using Credentials the first time, tried to use log in with OAuth later and vice versa. 
> 
> However, it's not super important to support it, we can drop it if required. 

Yeah, think we can leave it for later

**Review by homanp — CHANGES_REQUESTED** [review]

File: app/api/agents.py [hunk 2]
```diff
@@ -21,7 +21,14 @@
 @router.post("/agents", name="Create agent", description="Create a new agent")
 async def create_agent(body: Agent, token=Depends(JWTBearer())):
     """Agents endpoint"""
-    decoded = decodeJWT(token)
+    is_oauth_token = False
```
[the remark this answers is not in the document]
Comment (homanp) [reply app/api/agents.py:24]: @ankushbhardwxj did you have the time look at my comments?

[the remark this answers is not in the document]
Comment (ankushbhardwxj) [reply app/api/agents.py:24]: @homanp : yeah I did check this out, didn't get time to get to it, will push the changes soon.

**Review by homanp — COMMENTED** [review]
Hi @ankushbhardwxj did you test your implementation with listing/creating objects both for OAuth and email/password users?!

**Comment by ankushbhardwxj** [comment]
@homanp : yes, I am able to create and list agents, tools, prompts, documents, etc. Which object were you not able to create when you tested earlier ? Did I miss out anything ?

**Comment by ankushbhardwxj** [comment]
@homanp : I guess this PR is ready for another review. Do let me know for changes.

**Comment by homanp** [comment]
> @homanp : I guess this PR is ready for another review. Do let me know for changes. 

Thanks! Will test and merge. This PR is next ip! 😻

**Review by homanp — CHANGES_REQUESTED** [review]
I can't login using regular auth with email password.

**Comment by ankushbhardwxj** [comment]
@homanp sure, will take a look at it. Few questions just to clarify,
- Is this a fresh user ? Did you do normal sign up and then do the log in ?
- Did you do login with OAuth and then try to login using normal email and password ?

**Comment by homanp** [comment]
> @homanp sure, will take a look at it. Few questions just to clarify,
> 
> * Is this a fresh user ? Did you do normal sign up and then do the log in ?
> * Did you do login with OAuth and then try to login using normal email and password ?

Hmm, no this was my fault. Seems to work now! Doing some final testing and then will merge! Thanks!

**Comment by homanp** [comment]
I'm merging this together with #222

**Pull request closed without merging**